src/paradigma/heart_rate/heart_rate_estimation.py

```diff
@@ -0,0 +1,185 @@
+import numpy as np
+from typing import Tuple
+
+from paradigma.config import HeartRateExtractionConfig
+from paradigma.heart_rate.tfd import nonsep_gdtfd
+
+def assign_sqa_label(ppg_prob, config: HeartRateExtractionConfig, acc_label=None) -> np.ndarray:
+    """
+    Assigns a signal quality label to every individual data point.
+
+    Parameters:
+    - ppg_prob: numpy array of probabilities for PPG
+    - acc_label: numpy array of labels for accelerometer (optional)
+    - fs: Sampling frequency
+
+    Returns:
+    - sqa_label: numpy array of signal quality assessment labels
+    """
+    # Default _label to ones if not provided
+    if acc_label is None:
+        acc_label = np.ones(len(ppg_prob))
+
+
+    # Number of samples in an epoch
+    fs = config.sampling_frequency
+    samples_per_epoch = config.window_length_s * fs
+
+    # Calculate number of samples to shift for each epoch
+    samples_shift = config.window_step_size_s * fs
+    n_samples = int((len(ppg_prob) + config.window_overlap_s) * fs)
+    data_prob = np.zeros(n_samples)
+    data_label_imu = np.zeros(n_samples, dtype=np.int8)
+
+    for i in range(n_samples):
+        # Start and end indices for current epoch
+        start_idx = int(np.floor((i - (samples_per_epoch - samples_shift)) / fs))
+        end_idx = int(np.floor(i / fs))
+
+        # Correct for first and last epochs
+        if start_idx < 0:
+            start_idx = 0
+        if end_idx > len(ppg_prob):
+            end_idx = len(ppg_prob)
+
+        # Extract probabilities and labels for the current epoch
+        prob = ppg_prob[start_idx:end_idx+1]
+        label_imu = acc_label[start_idx:end_idx+1]
+
+        # Calculate mean probability and majority voting for labels
+        data_prob[i] = np.mean(prob)
+        data_label_imu[i] = int(np.mean(label_imu) >= 0.5)
+
+    # Set probability to zero if majority IMU label is 0
+    data_prob[data_label_imu == 0] = 0
+    sqa_label = (data_prob > config.threshold_sqa).astype(np.int8)
```

Comment:
Two questions:
1. Should the probability be (1) larger than, or (2) larger than or equal to?
2. Why do you convert it to `np.int8`? Why don't you store it as a boolean?

Reply:
1. Probability is larger than (1) instead of (2) in my Matlab code, however this is not entirely correct so I would want to have (2) so I adjust it here. I won't expect that this really matters for the results because I won't expect that many probabilities to be exactly the threshold

Reply:
Great! You might run into issues with storing as boolean due to TSDF constraints.

Reply:
This step doesn't have to be stored in tsdf luckily :). It's just an input for extracting the ppg segments suited for HR estimation